### PR TITLE
feat: WIP on amountFormat

### PR DESCRIPTION
**Pull request by Majorfi**
## Description
WIP implementation of a refactor on the amount display to get rid of the troubles of format format format format format format, have a react component to avoid having to set the suppress hydratation warning and display an information tooltip, handle edge case (big value, dust amount) and symbol displaying

## Related Issue
NA

## Motivation and Context
See description

## How Has This Been Tested?
Locally tested on the vault list page on Ethereum, still some work to do

### CODE REVIEW

File: apps/common/utils/format.tsx
Comment (karelianpie) [inline apps/common/utils/format.tsx:1]: We need to test this 😅 

Did you have a chance to see https://github.com/yearn/web-lib/pull/281

File: apps/common/utils/format.test.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+			T.formatLocalAmount(0.001, 2, 'OPT', T.defaultOptions).normalize('NFKC'),
+			'0,001 OPT',
+			'Amount above 0.01 should be formatted as is'
+		);
+		assert.equal(
+			T.formatLocalAmount(0.000000000000123, 18, 'YFI', T.defaultOptions).normalize('NFKC'),
+			'0,000000000000123 YFI',
+			'Amount should be formatted with the specified number of decimals'
+		);
+
```
Comment (karelianpie) [inline apps/common/utils/format.test.ts:102]: We need some edge cases, like when the number is too small to be formatted for example:

```suggestion
		assert.equal(
			T.formatLocalAmount(0.000000000000000000123, 18, 'YFI', T.defaultOptions).normalize('NFKC'),
			'0,00 YFI',
			'Amount should be formatted with the specified number of decimals'
		);
```

File: apps/common/utils/format.tsx [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	shouldDisplaySymbol: true,
+	shouldCompactValue: true
+};
+function assertValidNumber(value: number | undefined, defaultValue: number, label: string): number {
+	if (value === undefined) {
+		return defaultValue;
+	}
+	if (value < 0) {
+		console.warn(`formatAmount: ${label} should be positive.`);
+		return (defaultValue);
```
Comment (karelianpie) [inline apps/common/utils/format.tsx:31]: can we remove these parentheses from all the returns, not a thing in TS :)
```suggestion
		return defaultValue;
```

File: apps/common/utils/format.tsx [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+	}
+	if (value > 18) {
+		console.warn(`formatAmount: ${label} should be less than 18.`);
+		return (18);
+	}
+	if (Number.isNaN(value)) {
```
Comment (karelianpie) [inline apps/common/utils/format.tsx:37]: I would all the checks that return `defaultValue` to a single if statement

File: apps/common/utils/format.tsx [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	if (options.maximumFractionDigits < options.minimumFractionDigits) {
+		options.maximumFractionDigits = options.minimumFractionDigits;
+	}
+
+	if (options.shouldDisplaySymbol === undefined) {
+		options.shouldDisplaySymbol = true;
+	}
+	if (options.shouldCompactValue === undefined) {
+		options.shouldCompactValue = true;
+	}
```
Comment (karelianpie) [inline apps/common/utils/format.tsx:86]:
```suggestion
	options.shouldDisplaySymbol ??= true;
	options.shouldCompactValue ??= true;
```

File: apps/common/utils/format.tsx [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+			...intlOptions,
+			maximumFractionDigits: Math.max(decimals, intlOptions.maximumFractionDigits || decimals)
+		}).format(amount).replace('EUR', symbol));
+	}
+	return (
+		new Intl.NumberFormat([locale, 'en-US'], intlOptions)
+			.format(amount)
+			.replace('EUR', symbol)
+	);
+}
```
Comment (karelianpie) [inline apps/common/utils/format.tsx:171]: nit: DRY this a bit?

```suggestion
			return formatCurrencyWithPrecision({amount, maxFractionDigits: 8, intlOptions, locale, symbol});
		}
		if (amount > 0.000000000001) {
			return formatCurrencyWithPrecision({amount, maxFractionDigits: 12, intlOptions, locale, symbol});
		}
		return formatCurrencyWithPrecision({amount, maxFractionDigits: decimals, intlOptions, locale, symbol});
	}
	return (
		new Intl.NumberFormat([locale, 'en-US'], intlOptions)
			.format(amount)
			.replace('EUR', symbol)
	);
}

type TFormatCurrencyWithPrecision = {
	amount: number;
	maxFractionDigits: number;
	intlOptions: Intl.NumberFormatOptions;
	locale: string;
	symbol: string;
}

function formatCurrencyWithPrecision({amount, maxFractionDigits, intlOptions, locale, symbol}: TFormatCurrencyWithPrecision): string {
	return new Intl.NumberFormat([locale, 'en-US'], {
		...intlOptions,
		maximumFractionDigits: Math.max(maxFractionDigits, intlOptions.maximumFractionDigits || maxFractionDigits)
	}).format(amount).replace('EUR', symbol);
}
```

File: apps/common/utils/format.tsx [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+	** - If symbol is any other token, we will display as `123,79 USDC` or `USDC 123.79` (US)
+	**********************************************************************************************/
+	let locale = 'fr-FR';
+	if (typeof(navigator) !== 'undefined') {
+		locale = navigator.language || 'fr-FR';
+	}
+	const {shouldDisplaySymbol, shouldCompactValue, ...rest} = options;
+	const intlOptions: Intl.NumberFormatOptions = rest;
+	let isPercent = false;
+	if (symbol && symbol !== '' && shouldDisplaySymbol) {
```
Comment (karelianpie) [inline apps/common/utils/format.tsx:112]: We can simplify this as empty string is falsey
```suggestion
	if (symbol && shouldDisplaySymbol) {
```

File: apps/common/utils/format.tsx [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+				shouldDisplaySymbol: true,
+				shouldCompactValue: false
+			}
+		}));
+	}, [props]);
+
+	return (
+		<span
+			suppressHydrationWarning
+			className={isZero(props.value) ? '' : 'tooltip cursor-help underline decoration-neutral-600/30 decoration-dotted underline-offset-4 transition-opacity hover:decoration-neutral-600'}>
```
Comment (karelianpie) [inline apps/common/utils/format.tsx:206]:
```suggestion
			className={isZero(props.value) ? '' : 'tooltip underline decoration-neutral-600/30 decoration-dotted underline-offset-4 transition-opacity hover:decoration-neutral-600'}>
```

**Review by karelianpie — APPROVED** [review]

**Review by karelianpie — DISMISSED** [review]
